src/v/storage/segment_reader.h

```diff
@@ -176,4 +182,65 @@ class segment_reader {
 using segment_reader_ptr = ss::lw_shared_ptr<segment_reader>;
 
 std::ostream& operator<<(std::ostream&, segment_reader_ptr);
+
+/**
+ * Enables reading from a series of segments sequentially using a single data
+ * source. The first segment in the list is read starting from the start file
+ * position. The last segment in the list is read upto the end file position.
+ */
+class concat_segment_data_source_impl final : public ss::data_source_impl {
+public:
+    concat_segment_data_source_impl(
+      std::vector<ss::lw_shared_ptr<segment>> segments,
+      size_t start_pos,
+      size_t end_pos,
+      ss::io_priority_class priority_class);
+
+    /// Reads a buffer from the current underlying segment handle. Once the
+    /// segment is depleted, moves over to the next segment, until all segments
+    /// are finished.
+    ss::future<ss::temporary_buffer<char>> get() override;
+    ss::future<> close() override;
+
+private:
+    /**
+     * Switches over to the next segment, if the current segment is depleted.
+     */
+    ss::future<ss::input_stream<char>> next_stream();
+
+private:
+    using segment_seq = std::vector<ss::lw_shared_ptr<segment>>;
+    using segment_iter = segment_seq::iterator;
+    segment_seq _segments;
+    std::optional<ss::input_stream<char>> _current_stream;
+    std::optional<segment_reader_handle> _current_handle;
+    segment_iter _current_pos;
+
+    // The file position from where the first segment will be read.
+    size_t _start_pos;
+
+    // The file position upto which the last segment will be read.
+    size_t _end_pos;
+    ss::io_priority_class _priority_class;
+};
+
+/**
+ * Provides a stream_provider interface for reading from underlying segments in
+ * sequence.
+ */
+class concat_segment_reader_view final : public stream_provider {
```

Comment:
Nit: I would call this `concat_segment_reader_handle` to be consistent with `segment_reader_handle`.

Reply:
In the initial version of the code I actually called it something like handle and the stream provider was called `reader_handle`, but then I thought it would be a bit misleading, because this structure is a collection of handles and not a real handle (it _has_ a handle). 

But if it makes more sense to call it `_handle` for consistency let me know and I can change it.

 If the comment https://github.com/redpanda-data/redpanda/pull/6516#discussion_r980040040 by @Lazin can be addressed if I understand correctly we might be able to do away with the interface altogether.